node/pkg/ethereum/moonbeamimpl.go

```diff
@@ -0,0 +1,259 @@
+	return e.BaseEth.ParseLogMessagePublished(log)
+}
+
+type MoonbeamSubscription struct {
+	errOnce   sync.Once
+	err       chan error
+	quit      chan error
+	unsubDone chan struct{}
+}
+
+func (sub *MoonbeamSubscription) Err() <-chan error {
+	return sub.err
+}
+
+var errUnsubscribed = errors.New("unsubscribed")
+
+func (sub *MoonbeamSubscription) Unsubscribe() {
+	sub.errOnce.Do(func() {
+		select {
+		case sub.quit <- errUnsubscribed:
+			<-sub.unsubDone
+		case <-sub.unsubDone:
+		}
+		close(sub.err)
+	})
+}
+
+func (e *MoonbeamImpl) SubscribeForBlocks(ctx context.Context, sink chan<- *common.NewBlock) (ethereum.Subscription, error) {
+	if e.BaseEth.client == nil {
+		panic("client is not initialized!")
+	}
+	if e.mbRpcCon == nil {
+		panic("mbRpcCon is not initialized!")
+	}
+	if e.mbClient == nil {
+		panic("mbClient is not initialized!")
+	}	
+
+	latestBlockNumber, err := e.getLatestBlockNumber(ctx)
+	if err != nil {
+		return nil, err
+	}
+	currentBlockNumber := latestBlockNumber
+
+	const DELAY_IN_MS = 1000
+
+	timer := time.NewTimer(1 * time.Millisecond) // Start immediately.
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				for {
+					// See if the next block has been created yet.
+					if currentBlockNumber > latestBlockNumber {
+						latestBlockNumber, err = e.getLatestBlockNumber(ctx)
+						if err != nil {
+							e.logger.Error("failed to look up latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("block", currentBlockNumber), zap.Error(err))
+							break
+						}
+
+						if currentBlockNumber > latestBlockNumber {
+							e.logger.Info("waiting for new latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("current_block", currentBlockNumber), zap.Uint64("latest_block", latestBlockNumber))
+							break
+						}
+					}
+
+					// Fetch the hash every time, in case it changes due to a rollback.
+					blockHash, err := e.getBlockHash(ctx, currentBlockNumber)
+					if err != nil {
+						e.logger.Error("failed to look up hash for current block", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber), zap.Error(err))
+						break
+					}
+
+					e.logger.Info("checking to see if block is finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+						zap.Uint64("block", currentBlockNumber), zap.Stringer("hash", blockHash))
+					finalized, err := e.isBlockFinalized(ctx, blockHash.Hex())
+					if err != nil {
+						e.logger.Error("failed to see if block is finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber), zap.Error(err))
+						break
+					}
+					if !finalized {
+						break
+					}
+
+					e.logger.Info("block is now finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+						zap.Uint64("block", currentBlockNumber), zap.Stringer("hash", blockHash))
+
+					ev, err := e.getBlock(ctx, currentBlockNumber)
```

Comment:
If the `getBlockHash` call above instead returned a NewBlock, you could avoid this call.